apis/python/src/tiledbsoma/_tiledb_object.py

```diff
@@ -82,20 +82,17 @@ def open(
         del platform_config  # unused
         context = _validate_soma_tiledb_context(context)
         try:
-            handle = _tdb_handles._open_with_clib_wrapper(
-                uri, mode, context, tiledb_timestamp
-            )
+            handle = _tdb_handles.open(uri, mode, context, tiledb_timestamp)
         except SOMAError:
             handle = cls._wrapper_type.open(uri, mode, context, tiledb_timestamp)
         return cls(
-            handle,
+            handle,  # type: ignore
```

Comment:
@thetorpedodog I'm still having issues with typing for these. I get this error "apis/python/src/tiledbsoma/_tiledb_object.py:89: error: Argument 1 to "TileDBObject" has incompatible type "Wrapper[Any | Any | Any]"; expected "_WrapperType_co"  [arg-type]". I'm not sure what do add to get this to work.

Reply:
At this point, `_WrapperType_co` is essentially "resolved", so it refers specifically to the wrapper type of the given SOMA object.

Take this example:

```python
def returns_something() -> int | str:
  ...

_T_co = TypeVar("_T_co", bound=Union[int, str], covariant=True)
class Holder(Generic[_T_co]):
  _it: _T_co
  def __init__(self, obj: Optional[_T_co]):
    if obj is None:
      self._it = returns_something()
    else:
      self._it = obj

class IntHolder(Holder[int]):
  pass

class StrHolder(Holder[str]):
  pass
```

In this case, this use of `returns_something` is not valid, because `returns_something` may return either of an `int` or a `str`. So an `IntHolder` might get a `str`, or a `StrHolder` might get an `int`.

I think the right thing to do here is to add a generic entry for the specialized SOMA type we expect, and use *that* rather than the return of `_tdb_handles.open`.

Reply:
I think I partially understand but not really. I've been staring at this code and trying to follow it but confusing myself now. Is this on the right track? To make this valid you would modify the constructor to be:

```
_T_co = TypeVar("_T_co", bound=Union[int, str], covariant=True)
class Holder(Generic[_T_co]):
  _it: _T_co
  def __init__(self, obj: Optional[_T_co]):
    if obj is None:
      self._it = obj

class IntHolder(Holder[int]):
  pass

class StrHolder(Holder[str]):
  pass
```

And then you would use it like this which would create you either the `IntHolder` or `StringHolder`. You wouldn't actually use `returns_something()` inside `Holder.__init__()`?
```
holder = Holder(returns_something())
```

Reply:
I probably should have mentioned the fact that in the example, Holder is a purely abstract class (like TileDBObject), and IntHolder/StrHolder are the concrete types which provide additional functionality (like, e.g., Collection or DataFrame). So calling `Holder(returns_something())` by itself wouldn't make sense, but its type would end up being `Holder[int | str]`, which is equivalent to `Holder[int] | Holder[str]` due to covariance.

The important thing I was trying to get at is that `returns_something` is like `_tdb_handles.open`.

There are a couple of ways we could resolve this. Both involve putting the held class in the concrete types:

```python
_T_co = TypeVar("_T_co", bound=Union[int, str], covariant=True)
class Holder(Generic[_T_co]):
  _CLS: ClassVar[Type[_T_co]]
  _it: _T_co

  @classmethod
  def open(cls, obj: Optional[_T_co]) -> Self:
    ...

class IntHolder(Holder[int]):
  _CLS = int
  ...

class StrHolder(Holder[str]):
  _CLS = str
  ...
```

Either we check the type of the thing returned from `returns_something`:

```python
class Holder(Generic[_T_co]):
  ...
  @classmethod
  def open(cls, obj: Optional[_T_co]) -> Self:
    if obj is None:
      maybe_obj = returns_something()
      if not isinstance(maybe_obj, cls._CLS):
        raise TypeErrror("something went wrong")
      obj = maybe_obj
    self._it = obj
```

or we use a property of the type we passed in to return the value we want:

```python
class Holder(Generic[_T_co]):
  ...
  @classmethod
  def open(cls, obj: Optional[_T_co]) -> Self:
    # This isn't a perfect example since there aren't any classmethods on both
    # int and str which return their own type.
    self._it = cls(0) if obj is None else obj
```

Since C++ doesn’t really have an equivalent to this type of class method (i.e., you can’t define static `SOMAObject::open` and have it work as `SOMADataFrame::open` and `SOMASparseNDArray::open` etc.), I think we’re left with the first version, where we check the type that the open operation returned.

This means that in the end, we probably want something like:

```python
# AnyWrapper is renamed to AnyRWWrapper.
_RWWrapper_co = TypeVar("_RWWrapper_co", bound=_tdb_handles.AnyRWWrapper, covariant=True)
# The new AnyWrapper includes wrappers for both the read-only and read-write handles.
_ROWrapper_co = TypeVar("_ROWrapper_co", bound=_tdb_handles.AnyWrapper, covariant=True)
class TileDBObject(somacore.SOMAObject, Generic[_RWWrapper_co, _ROWrapper_co]):
  _rw_wrapper_type: Type[_RWWrapper_co]
  _ro_wrapper_type: Type[_ROWrapper_co]  # these should maybe be ClassVars

  _handle: Union[_RWWrapper_co, _ROWrapper_co]

  def __init__(self, handle: Union[_RWWrapper_co, _ROWrapper_co], ...) -> None:
    ...
    self._handle = handle

  @classmethod
  def open(cls, ...) -> Self:
    try:
      handle = _tdb_handles.open(...)
      if not isinstance(handle, cls._ro_wrapper_type):
        raise WhateverError(...)
      # Now we're guaranteed that `handle` is a `_ROWrapper_co` instance.
    except ...:
      handle = self._rw_wrapper_type.open(...)
    return cls(handle, ...)
```

Reply:
Thanks so much for explaining! I have something that works for this now.